### DISCUSSION TITLE
How can I use xtts for Turkish speech synthesis?

### DISCUSSION THREAD

**Discussion by trappedinspacetime** [coqui-ai/TTS#3312] [General Q&A]
I used to use the following line for Turkish tts:

    tts --text 'on dokuz elli iki ' --model_name tts_models/tr/common-voice/glow-tts --out_path result.mp3
    
But the above model is poor quality for a Turkish tts.

I found the below new voice cloning xtts line which works very well:

        tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 --text "Bugün okula gitmek istemiyorum." --speaker_wav untitled.wav --language_idx tr

How can I use it for only text to speech in Turkish?

I tried something like :

     tts --text 'on dokuz elli iki' --model_name tts_models/multilingual/multi-dataset/xtts_v2 --out_path result.mp3 --speaker_idx tr

But it gives the following error:
      > Using model: xtts
 > Text: on dokuz elli iki
 > Text splitted to sentences.
['on dokuz elli iki']
Traceback (most recent call last):
  File "/home/kenn/.pyenv/versions/3.10.7/bin/tts", line 8, in <module>
    sys.exit(main())
  File "/home/kenn/.pyenv/versions/3.10.7/lib/python3.10/site-packages/TTS/bin/synthesize.py", line 538, in main
    wav = synthesizer.tts(
  File "/home/kenn/.pyenv/versions/3.10.7/lib/python3.10/site-packages/TTS/utils/synthesizer.py", line 325, in tts
    raise ValueError(
ValueError:  [!] Missing speakers.json file path for selecting speaker tr.Define path for speaker.json if it is a multi-speaker model or remove defined speaker idx.

**Comment by erturkkadir** [comment]
@trappedinspacetime the following works fine here;

    tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 --text "Bugün okula gitmek istemiyorum." --speaker_wav untitled.wav --language_idx tr

However, if I start server with 
```
python3.10 TTS/server/server.py --use_cuda true --port 7865 --model_name tts_models/multilingual/multi-dataset/xtts_v2
```

then app crashed with following error;

 File "/var/www/html/coqui/TTS/TTS/server/server.py", line 104, in <module>
    synthesizer = Synthesizer(
  File "/var/www/html/coqui/TTS/TTS/utils/synthesizer.py", line 93, in __init__
    self._load_tts(tts_checkpoint, tts_config_path, use_cuda)
  File "/var/www/html/coqui/TTS/TTS/utils/synthesizer.py", line 183, in _load_tts
    self.tts_config = load_config(tts_config_path)
  File "/var/www/html/coqui/TTS/TTS/config/__init__.py", line 82, in load_config
    ext = os.path.splitext(config_path)[1]
  File "/usr/lib/python3.10/posixpath.py", line 118, in splitext
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType


one more thing, I was able to get response from other model (tts_models/tr/common-voice/glow-tts) with following cURL
 curl -X POST XXX.XXX.XXX/.XXX:XXX/api/tts --header "language-id: tur" --header "style-wav: referans.wav" --header "speaker-id: tur" -d "text=okula gitmek istemiyorum&language-id=tur" --output snd.wav

Hope you found a good solution

**Reply by trappedinspacetime** [reply]
Unfortunately, no. I hope somebody finds a solution.

**Reply by eginhard** [reply]
In our fork the server now works with XTTS: https://github.com/idiap/coqui-ai-TTS

**Comment by Rumeysakeskin** [comment]
@trappedinspacetime , If you want to use tts for a specific language,

1. You can check all Coqui available speakers with the following command:
```
!pip install TTS
!tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 \
    --list_speaker_idx
```
2. You can do inference using one of the available speakers using the following command:

```
!tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 \
     --text "Bugün okula gitmek istemiyorum." \
     --speaker_idx "Asya Anara" \
     --language_idx tr \
     --use_cuda true
```

```
import IPython
IPython.display.Audio("/content/tts_output.wav")
```

**Reply by trappedinspacetime** [reply]
@Rumeysakeskin 

        tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 \
             --text "Bugün okula gitmek istemiyorum." \
             --speaker_idx "Asya Anara" \
             --language_idx tr 
        tts_models/multilingual/multi-dataset/xtts_v2 is already downloaded.
        Using model: xtts
        Text: Bugün okula gitmek istemiyorum.
        Text split into sentences.
        Input: ['Bugün okula gitmek istemiyorum.']
        Traceback (most recent call last):
          File "/home/kenn/.pyenv/versions/3.12.2/bin/tts", line 8, in <module>
            sys.exit(main())
                     ^^^^^^
          File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/bin/synthesize.py", line 413, in main
            api.tts_to_file(
          File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/api.py", line 377, in tts_to_file
            wav = self.tts(
                  ^^^^^^^^^
          File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/api.py", line 323, in tts
            wav = self.synthesizer.tts(
                  ^^^^^^^^^^^^^^^^^^^^^
          File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/utils/synthesizer.py", line 366, in tts
            raise ValueError(
        ValueError:  [!] Missing speakers.json file path for selecting speaker Asya Anara.Define path for speaker.json if it is a multi-speaker model or remove defined speaker idx.

**Reply by Rumeysakeskin** [reply]
I guess the system cannot find the speaker "Asya Anara" you're trying to use.
You can check all Coqui available speakers with the following command:
```
 tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 --list_speaker_idx
```

**Reply by eginhard** [reply]
@trappedinspacetime The `speakers_xtts.pth` might be missing from your XTTS model folder. If necessary, you can download it from https://huggingface.co/coqui/XTTS-v2/tree/main

**Reply by trappedinspacetime** [reply]
@Rumeysakeskin @eginhard  Thank you for your help. I was late in writing due to health issues, I just realized what you wrote. 
Since tts in this repo has not been updated and Python 3.12.2 is not supported, I installed the new version as follows: pip install coqui-tts -U
I ran the command you said. I got this result:
tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 --list_speaker_idx
tts_models/multilingual/multi-dataset/xtts_v2 is already downloaded.
Using model: xtts
Model only has a single speaker.

By the way, I used this repo "https://huggingface.co/coqui/XTTS-v2"   from HuggingFace to my desktop. But I think TTS may be using the models left in the old installed computer. I used the following commands to prevent this, but I got stuck there too. Now I will share it with you.

          tts --text "Merhaba Dünya "     --model_path /home/kenn/Desktop/2025-04/XTTS-v2/model.pth     --config_path /home/kenn/Desktop/2025-04/XTTS-v2/config.json     --out_path /home/kenn/Desktop/2025-04/XTTS-v2/speech.wav
          Using model: xtts
          Traceback (most recent call last):
            File "/home/kenn/.pyenv/versions/3.12.2/bin/tts", line 8, in <module>
              sys.exit(main())
                       ^^^^^^
            File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/bin/synthesize.py", line 363, in main
              api = TTS(
                    ^^^^
            File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/api.py", line 103, in __init__
              self.load_tts_model_by_path(model_path, config_path, gpu=gpu)
            File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/api.py", line 257, in load_tts_model_by_path
              self.synthesizer = Synthesizer(
                                 ^^^^^^^^^^^^
            File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/utils/synthesizer.py", line 99, in __init__
              self._load_tts(self.tts_checkpoint, self.tts_config_path, use_cuda)
            File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/utils/synthesizer.py", line 215, in _load_tts
              self.tts_model.load_checkpoint(self.tts_config, tts_checkpoint, eval=True)
            File "/home/kenn/.pyenv/versions/3.12.2/lib/python3.12/site-packages/TTS/tts/models/xtts.py", line 745, in load_checkpoint
              raise ValueError(msg)
          ValueError: You passed a file to `checkpoint_dir=`. Use `checkpoint_path=/home/kenn/Desktop/2025-04/XTTS-v2/model.pth` instead.

But when I follow the suggestion in the error message, I get this error:

    tts --text "Merhaba Dünya "     --checkpoint_path=/home/kenn/Desktop/2025-04/XTTS-v2/model.pth     --config_path /home/kenn/Desktop/2025-04/XTTS-v2/config.json     --out_path /home/kenn/Desktop/2025-04/XTTS-v2/speech.wav
    usage: tts [-h] [--list_models] [--model_info_by_idx MODEL_INFO_BY_IDX]
               [--model_info_by_name MODEL_INFO_BY_NAME] [--text TEXT]
               [--model_name MODEL_NAME] [--vocoder_name VOCODER_NAME]
               [--config_path CONFIG_PATH] [--model_path MODEL_PATH]
               [--out_path OUT_PATH] [--use_cuda] [--device DEVICE]
               [--vocoder_path VOCODER_PATH]
               [--vocoder_config_path VOCODER_CONFIG_PATH]
               [--encoder_path ENCODER_PATH]
               [--encoder_config_path ENCODER_CONFIG_PATH] [--pipe_out]
               [--speakers_file_path SPEAKERS_FILE_PATH]
               [--language_ids_file_path LANGUAGE_IDS_FILE_PATH]
               [--speaker_idx SPEAKER_IDX] [--language_idx LANGUAGE_IDX]
               [--speaker_wav SPEAKER_WAV [SPEAKER_WAV ...]]
               [--gst_style GST_STYLE]
               [--capacitron_style_wav CAPACITRON_STYLE_WAV]
               [--capacitron_style_text CAPACITRON_STYLE_TEXT]
               [--list_speaker_idxs] [--list_language_idxs]
               [--reference_wav REFERENCE_WAV]
               [--reference_speaker_idx REFERENCE_SPEAKER_IDX]
               [--progress_bar | --no-progress_bar] [--source_wav SOURCE_WAV]
               [--target_wav [TARGET_WAV ...]] [--voice_dir VOICE_DIR]
    tts: error: unrecognized arguments: --checkpoint_path=/home/kenn/Desktop/2025-04/XTTS-v2/model.pth

**Reply by eginhard** [reply]
Yeah, the error message is a bit misleading here, but you need to use: `tts --text "Merhaba Dünya " --model_path /home/kenn/Desktop/2025-04/XTTS-v2 --config_path /home/kenn/Desktop/2025-04/XTTS-v2/config.json --out_path /home/kenn/Desktop/2025-04/XTTS-v2/speech.wav`

But if you didn't finetune the model you can also just do the following and let Coqui handle the download for you: `tts --model_name tts_models/multilingual/multi-dataset/xtts_v2 --out_path /home/kenn/Desktop/2025-04/XTTS-v2/speech.wav`

In either case you'll also have to add either a `--speaker_idx` or a `--speaker_wav` argument for it to work.

**Reply by trappedinspacetime** [reply]
@eginhard Thanks for the help. After running the command below, I was able to convert text to speech. I have another question. When we use Python, is there a method where we can load the model into memory and get faster results from there in instant translations? After entering the command below, it took more than half a minute for the result to appear.

        tts --text "Merhaba Dünya " --model_path /home/kenn/Desktop/2025-04/XTTS-v2 --config_path /home/kenn/Desktop/2025-04/XTTS-v2/config.json --out_path /home/kenn/Desktop/2025-04/XTTS-v2/speech.wav --speaker_idx "Asya Anara"     --language_idx "tr"
        Using model: xtts
        Text: Merhaba Dünya 
        Text split into sentences.
        Input: ['Merhaba Dünya']
        Processing time: 3.816
        Real-time factor: 2.298
        Saved TTS output to /home/kenn/Desktop/2025-04/XTTS-v2/speech.wav

**Reply by eginhard** [reply]
Yes, see https://coqui-tts.readthedocs.io/en/latest/models/xtts.html#tts-api
